src/charm.py

```diff
@@ -395,6 +403,98 @@ def _on_vault_pki_certificate_creation_request(
             event.certificate_signing_request, event.relation_id
         )
 
+    def _on_create_backup_action(self, event: ActionEvent) -> None:
+        """Handle the create-backup action.
+
+        Creates a snapshot and stores it on S3 storage.
+        Outputs the ID of the backup to the user.
+
+        Args:
+            event: ActionEvent
+        """
+        if not self.unit.is_leader():
+            event.fail(message="Only leader unit can perform backup operations.")
+            return
+
+        if not self._is_relation_created(S3_RELATION_NAME):
+            event.fail(message="Failed to perform backup. S3 relation not created.")
+            return
+
+        if missing_parameters:= self._get_missing_s3_parameters():
+            event.fail(message=f"Failed to perform backup. S3 parameters missing: {missing_parameters}")
+            return
+
+        s3_parameters = self._get_s3_parameters()
+
+        try:
+            s3 = S3(
+                access_key=s3_parameters["access-key"],
+                secret_key=s3_parameters["secret-key"],
+                endpoint=s3_parameters["endpoint"],
+                region=s3_parameters.get("region"),
+            )
+        except S3Error:
```

Comment:
Should we log the error? (it's not logged where it's raised either, but here would be a better place anyway)

Reply:
Done